### PR TITLE
px4_platform_common: atomic support larger types with critical sections (on NuttX)

### PR DESCRIPTION
**Pull request by dagar**
This extends px4::atomic to use a critical section on NuttX if the size exceeds what can be done with lock-free atomic instructions. The intent it to use this in drivers to get the sample timestamp from the interrupt without littering ifdefs everywhere, although I'm open to alternatives. Later this could be extended to use locks for larger sizes on other platforms like std::atomic (or just use std::atomic directly).

### CODE REVIEW

File: platforms/common/include/px4_platform_common/atomic.h [hunk 1]
```diff
@@ -66,11 +70,11 @@ class atomic
 {
 public:
 
-#ifdef __PX4_NUTTX
+#if defined(__PX4_POSIX)
```
Comment (bkueng) [inline platforms/common/include/px4_platform_common/atomic.h:73]:
```suggestion
#if !defined(__PX4_NUTTX)
```

**Review by bkueng — APPROVED** [review]
You could also use a template specialization, that should be a bit less cluttered with ifdef's.

File: platforms/common/include/px4_platform_common/atomic.h (hunk 1 above)
Comment (dagar) [reply platforms/common/include/px4_platform_common/atomic.h:73]: It's also skipped on QuRT where the `_value` is simply volatile.

**Comment by dagar** [comment]
> You could also use a template specialization, that should be a bit less cluttered with ifdef's.

Without std::enable_if I wasn't sure how to do it by size without making a mess. Any suggestions? I'm also assuming we'll come back to this for the other platforms eventually.

File: platforms/common/include/px4_platform_common/atomic.h (hunk 1 above)
Comment (bkueng) [reply platforms/common/include/px4_platform_common/atomic.h:73]: Alright. I slightly prefer the explicit exclusion list to avoid surprises.

**Comment by bkueng** [comment]
I was thinking something like this:
```
#if defined(__PX4_NUTTX)
template <>
class atomic<hrt_abstime>
{
...
```
Assuming `hrt_abstime` is the only type we need.

**Pull request merged**